packages/wrangler/src/__tests__/secret.test.ts

```
@@ -44,83 +48,180 @@ describe("wrangler secret", () => {
-      expect(std.err).toMatchInlineSnapshot(`
-        "Missing script name
+    describe("non-interactive", () => {
+      const mockStdIn = useMockStdin({ isTTY: false });
 
-        [32m%s[0m If you think this is a bug then please create an issue at https://github.com/cloudflare/wrangler2/issues/new."
-      `);
-      expect(error).toMatchInlineSnapshot(`[Error: Missing script name]`);
+      it("should create a secret, from piped input", async () => {
+        mockPutRequest({ name: "the-key", text: "the-secret" });
+        // Pipe the secret in as three chunks to test that we reconstitute it correctly.
+        mockStdIn.send("the", "-", "secret");
+        await runWrangler("secret put the-key --name script-name");
+
+        expect(std.out).toMatchInlineSnapshot(`
+          "🌀 Creating the secret for script script-name
+          ✨ Success! Uploaded secret the-key"
+        `);
+        expect(std.warn).toMatchInlineSnapshot(`""`);
+        expect(std.err).toMatchInlineSnapshot(`""`);
+      });
+
+      it("should error if the piped input fails", async () => {
+        mockPutRequest({ name: "the-key", text: "the-secret" });
+        mockStdIn.throwError(new Error("Error in stdin stream"));
+        await expect(
+          runWrangler("secret put the-key --name script-name")
+        ).rejects.toThrowErrorMatchingInlineSnapshot(`"Error in stdin stream"`);
+
+        expect(std.out).toMatchInlineSnapshot(`""`);
+        expect(std.warn).toMatchInlineSnapshot(`""`);
+      });
+
+      describe("with accountId", () => {
+        mockAccountId({ accountId: "" });
+
+        it("should error if a user has no account", async () => {
+          await expect(
+            runWrangler("secret put the-key --name script-name")
+          ).rejects.toThrowErrorMatchingInlineSnapshot(
+            `"No account id found, quitting..."`
+          );
+        });
+
+        it("should use the account from wrangler.toml", async () => {
```

why did this test not require mockPutRequest?

---

Good question! The `PUT` request for the available accounts is not made because we have already got an `account_id` from the `wrangler.tom`.

---

but the mock is for the PUT call that sets the secret?

---

OMG! Even better question...

---

It turns out that `submitSecret()` is catching the error and swallowing it!
Great spot!

---

omg

---

And fixing that showed up a bunch of other test failures!